snuba/snuba_migrations/generic_metrics/0031_counters_meta_table_mv.py

```diff
@@ -0,0 +1,71 @@
+from typing import Sequence
+
+from snuba.clickhouse.columns import AggregateFunction, Column, DateTime, String, UInt
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.migrations.operations import OperationTarget
+from snuba.utils.schemas import Float
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+    view_name = "generic_metric_counters_meta_aggregation_mv"
+    dest_table_name = "generic_metric_counters_meta_aggregated_local"
+    dest_table_columns: Sequence[Column[Modifiers]] = [
+        Column("org_id", UInt(64)),
+        Column("project_id", UInt(64)),
+        Column("use_case_id", String(Modifiers(low_cardinality=True))),
+        Column("metric_id", UInt(64)),
+        Column("tag_key", String()),
+        Column("tag_value", String()),
+        Column("timestamp", DateTime(modifiers=Modifiers(codecs=["DoubleDelta"]))),
+        Column("retention_days", UInt(16)),
+        Column("value", AggregateFunction("sum", [Float(64)])),
+    ]
+    storage_set_key = StorageSetKey.GENERIC_METRICS_COUNTERS
+
+    def forwards_ops(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateMaterializedView(
+                storage_set=self.storage_set_key,
+                view_name=self.view_name,
+                columns=self.dest_table_columns,
+                destination_table_name=self.dest_table_name,
+                target=OperationTarget.LOCAL,
+                query="""
+                SELECT
+                    org_id,
+                    project_id,
+                    use_case_id,
+                    metric_id,
+                    tag_key,
+                    tag_value,
+                    toStartOfWeek(timestamp) as timestamp,
+                    retention_days,
+                    sumState(count_value) as count
+                FROM generic_metric_counters_raw_local
+                ARRAY JOIN
+                    tags.key AS tag_key,
+                    tags.raw_value AS tag_value
```

Comment:
I want to make sure that I understand how this ARRAY JOIN works and that we think it's the right behavior.

If I insert:
```
{org_id=1,project_id=1,use_case_id="custom",metric_id=2,
  tags.key=["foo1","foo2"],tags.value=["bar1","bar2"]}
{org_id=1,project_id=1,use_case_id="custom",metric_id=2,
  tags.key=["foo1","foo2"],tags.value=["bar2","bar3"]}
{org_id=1,project_id=1,use_case_id="custom",metric_id=2,
  tags.key=["foo1","foo2"],tags.value=["bar3","bar4"]}
```

what rows should be output? I assume something like (assuming all prior columns are equal)

```
tags_key,tag_value,count
foo1,bar1,1
foo2,bar2,1
foo1,bar2,1
foo2,bar3,1
foo1,bar3,1
foo2,bar4,1
```

my question would be, do we actually want to store unique counts for key=value pairs, or should we be counting the number of times a unique string shows up and some context e.g.:

```
user_value,type,count
foo1,key,3
foo2,key,3
bar1,value,1
bar2,value,2
bar3,value,2
bar4,value,1
```

This drops information about the relationship between tag keys and values but I could imagine it being a much more efficient encoding scheme for some kinds of input data.

Reply:
> I assume something like (assuming all prior columns are equal)

You are right about this.

> This drops information about the relationship between tag keys and values
We need the relationship between keys and values, because one of the two autocomplete queries is "given this tag key, tell me it's values".

Reply:
sounds good to me

Reply:
> This drops information about the relationship between tag keys and values
We need the relationship between keys and values, because one of the two autocomplete queries is "given this tag key, tell me it's values".

Maybe I might be missing some context. Would it be better to have the relationship between a tag key and its tag values (whose relationship is 1 -> N) be represented like this in the aggregate table?
```
tag_keys int
tag_values Array(str)
```
Storing data this way would make sure that the number of rows in the aggregate table is restricted only by the number of unique metric_id, tag_keys combination. Compared to the current approach of unique metric_id, tag_keys, tag_values combination. 

To answer the question about `give me all tag values for a given metric and tag key` return the full array back to the end user.